### PR TITLE
Ping pong

### PR DESCRIPTION
**Pull request by notkriswagner**
Added ping pong test case and fixed bugs in ping pong code. Also did some examples refactoring and added error struct/functions to use across the whole library.

### CODE REVIEW

File: libsnowflakeclient/examples/CMakeLists.txt [hunk 1]
```diff
@@ -3,8 +3,43 @@
...
         "../include"
+        "utils"
 )
+set(SOURCE_UTILS
+        example_setup.c)
 
-add_subdirectory(connect)
-add_subdirectory(select1)
-add_subdirectory(transaction)
+add_executable(ex_connect ${SOURCE_UTILS} connect.c)
```
Comment (smtakeda) [inline libsnowflakeclient/examples/CMakeLists.txt]: Can you please explain what the advantage of merging CMakeLists.txt here is?

File: libsnowflakeclient/CMakeLists.txt [hunk 2]
```diff
@@ -21,7 +21,10 @@ set(SOURCE_FILES
         lib/connection.c
         lib/results.c
         lib/uuid4.c
-        lib/basic_types.c)
+        lib/basic_types.c
+        lib/error.h
```
Comment (smtakeda) [inline libsnowflakeclient/CMakeLists.txt]: I don't think you need to include header file in SOURCE_FILES.

File: libsnowflakeclient/CMakeLists.txt [hunk 3]
```diff
@@ -21,7 +21,10 @@ set(SOURCE_FILES
...
+        lib/error.c
+        lib/snowflake_client_int.h)
```
Comment (smtakeda) [inline libsnowflakeclient/CMakeLists.txt]: Same here

File: libsnowflakeclient/lib/error.c [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+                                  int line) {
+    size_t sfqid_size = 1;
+    // NULL error passed in. Should never happen
+    if (!error) {
+        return NULL;
+    }
+
+    // Allocate enough memory for sfqid
+    sfqid_size += strlen(sfqid);
+    error->sfqid = (char *) SF_CALLOC(1, sfqid_size);
```
Comment (smtakeda) [inline libsnowflakeclient/lib/error.c]: probably we need to revisit this. If sfqid is given, you don't need to copy here but pass the pointer to sfqid. This way, you won't need to free memory for sfqid.

**Comment by smtakeda** [comment]
Overall looks good for now.

File: libsnowflakeclient/lib/error.c (hunk 4 above)
Comment (jianshenghuang) [reply libsnowflakeclient/lib/error.c]: yeah we have to be careful with memory allocation in C so it doesn't leak.

Comment (notkriswagner) [reply libsnowflakeclient/lib/error.c]: I've changed this to be a fixed size char array.

File: libsnowflakeclient/examples/CMakeLists.txt (hunk 1 above)
Comment (notkriswagner) [reply libsnowflakeclient/examples/CMakeLists.txt]: In a future pull request, this file will use a foreach loop to reduce the amount of code repeated. This way we'll just have to add the name of the example and everything will be done for us.

File: libsnowflakeclient/CMakeLists.txt (hunk 3 above)
Comment (notkriswagner) [reply libsnowflakeclient/CMakeLists.txt]: The error header file I can remove. For snowflake_client_int.h, the entire definition exists in the header file (and there is no source file) so I need to include that as a "source file" to get this to work.

Comment (smtakeda) [reply libsnowflakeclient/CMakeLists.txt]: hmm, ok.

**Comment by notkriswagner** [comment]
This fixes snowflakedb/pdo_snowflake#5

**Pull request merged**